### PR TITLE
Initial port of TMK's ibmpc_usb converter code

### PR DESCRIPTION
**Pull request by marfrit**
## Description

I ported the ibmpc_usb converter from TMK. 
hasu is combining the ps2_usb, terminal_usb and xt_usb converters in ibmpc_usb - and so should qmk.
Not entirely sure about the location of ibmpc.c and ibmpc.h, since they reside in /tmk_core/protocol/ in TMK.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* https://github.com/qmk/qmk_firmware/issues/7321 (partially) 

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] 0I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/converter/ibm_terminal/keymaps/default/rules.mk [hunk 1]
```diff
@@ -1,20 +1,21 @@
...
 #   the appropriate keymap folder that will get included automatically
 #
 BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
-MOUSEKEY_ENABLE = yes        # Mouse keys
+MOUSEKEY_ENABLE = no        # Mouse keys
 EXTRAKEY_ENABLE = yes       # Audio control and System control
-CONSOLE_ENABLE = no         # Console for debug
-COMMAND_ENABLE = no         # Commands for debug and configuration
+CONSOLE_ENABLE = yes        # Console for debug
+COMMAND_ENABLE = yes        # Commands for debug and configuration
```
Comment (drashna) [inline keyboards/converter/ibm_terminal/keymaps/default/rules.mk:9]: Honestly, I'm not a fan of these changes. 

Mousekeys may be more useful to people.  And console and command are really mostly only useful for debugging.  

This change (and the matching one to the keyboard level rules.mk) should be reverted.

File: keyboards/converter/ibmpc_usb/config.h [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef CONFIG_H
+#define CONFIG_H
```
Comment (drashna) [inline keyboards/converter/ibmpc_usb/config.h:20]:
```suggestion
#pragma once
```

File: keyboards/converter/ibmpc_usb/config.h [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+
+#define VENDOR_ID       0xFEED
+#define PRODUCT_ID      0x6536
+#define DEVICE_VER      0x0100
+#define MANUFACTURER    QMK
+#define PRODUCT         PS/2 Keyboard
+#define DESCRIPTION     convert IBM PC keyboard to USB
```
Comment (drashna) [inline keyboards/converter/ibmpc_usb/config.h:28]:
```suggestion
```

File: keyboards/converter/ibmpc_usb/config.h [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    EIMSK |= (1<<INT1);         \
+} while (0)
+
+#define IBMPC_INT_OFF() do {    \
+    EIMSK &= ~(1<<INT1);        \
+} while (0)
+
+#define IBMPC_INT_VECT    INT1_vect
+
+#endif
```
Comment (drashna) [inline keyboards/converter/ibmpc_usb/config.h:95]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/converter/ibm_terminal/keymaps/default/rules.mk (hunk 1 above)
Comment (marfrit) [reply keyboards/converter/ibm_terminal/keymaps/default/rules.mk:9]: These got in by accident, I wasn't meant to change this

**Comment by marfrit** [comment]
Hi, 
I think I got everything mentioned. Changes to ibm_terminal were not intended and I reverted those.

File: keyboards/converter/ibmpc_usb/rules.mk [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+MCU = atmega32u4
+
+# Bootloader selection
+#   Teensy       halfkay
+#   Pro Micro    caterina
+#   Atmel DFU    atmel-dfu
+#   LUFA DFU     lufa-dfu
+#   QMK DFU      qmk-dfu
+#   ATmega32A    bootloadHID
+#   ATmega328P   USBasp
```
Comment (drashna) [inline keyboards/converter/ibmpc_usb/rules.mk:11]:
```suggestion
```

**Review by drashna — APPROVED** [review]

File: keyboards/40percentclub/foobar/keymaps/qfoobar/config.h [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#define PERMISSIVE_HOLD
```
Comment (fauxpark) [inline keyboards/40percentclub/foobar/keymaps/qfoobar/config.h:1]:
```suggestion
#pragma once

#define PERMISSIVE_HOLD
```

File: keyboards/40percentclub/foobar/keymaps/qfoobar/rules.mk [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+#COMMAND_ENABLE = no          # Commands for debug and configuration
+#NKRO_ENABLE = no             # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality
+MIDI_ENABLE = no             # MIDI controls
+AUDIO_ENABLE = no            # Audio output on port C6
+UNICODE_ENABLE = yes          # Unicode
+BLUETOOTH_ENABLE = no        # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = yes        # Enable WS2812 RGB underlight.
+#API_SYSEX_ENABLE = no
+
```
Comment (fauxpark) [inline keyboards/40percentclub/foobar/keymaps/qfoobar/rules.mk:16]:
```suggestion
BOOTMAGIC_ENABLE = no
MOUSEKEY_ENABLE = no
UNICODE_ENABLE = yes
RGBLIGHT_ENABLE = yes
LTO_ENABLE = yes
```

File: keyboards/converter/ibm_terminal/keymaps/default/rules.mk [hunk 8]
```diff
@@ -1,20 +1,21 @@
...
 NKRO_ENABLE = yes           # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
 BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
 MIDI_ENABLE = no            # MIDI controls
 AUDIO_ENABLE = no           # Audio output on port C6
 UNICODE_ENABLE = no         # Unicode
 BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
 RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
```
Comment (fauxpark) [inline keyboards/converter/ibm_terminal/keymaps/default/rules.mk:16]:
```suggestion
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/mitosis/keymaps/qmitosis/config.h [hunk 9]
```diff
@@ -0,0 +1,4 @@
+#define PERMISSIVE_HOLD
```
Comment (fauxpark) [inline keyboards/mitosis/keymaps/qmitosis/config.h:1]:
```suggestion
#pragma once

#define PERMISSIVE_HOLD
```

File: keyboards/converter/usb_usb/keymaps/marhasu/config.h [hunk 10]
```diff
@@ -0,0 +1,2 @@
+#define PERMISSIVE_HOLD
```
Comment (fauxpark) [inline keyboards/converter/usb_usb/keymaps/marhasu/config.h:1]:
```suggestion
#pragma once

#define PERMISSIVE_HOLD
```

File: keyboards/converter/ibmpc_usb/keymaps/default/config.h [hunk 11]
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/keymaps/default/config.h:1]: This file can be deleted.

File: keyboards/converter/ibmpc_usb/ibmpc_usb.h [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef IBMPC_USB_H
+#define IBMPC_USB_H
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/ibmpc_usb.h:19]:
```suggestion
#pragma once
```

File: keyboards/converter/ibmpc_usb/ibmpc_usb.h [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+     kind == PC_AT ? "AT" :   \
+     kind == PC_TERMINAL ? "TERMINAL" :   \
+     "NONE")
+
+extern uint16_t keyboard_id;
+extern keyboard_kind_t keyboard_kind;
+extern const uint8_t map_cs1[MATRIX_ROWS][MATRIX_COLS];
+extern const uint8_t map_cs2[MATRIX_ROWS][MATRIX_COLS];
+extern const uint8_t map_cs3[MATRIX_ROWS][MATRIX_COLS];
+#endif
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/ibmpc_usb.h:83]:
```suggestion
```

File: keyboards/converter/ibmpc_usb/ibmpc.h [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+*/
+
+#ifndef IBMPC_H
+#define IBMPC_H
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/ibmpc.h:39]:
```suggestion
#pragma once
```

File: keyboards/converter/ibmpc_usb/ibmpc.h [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+    data_hi();
+}
+
+/* inhibit device to send(XT) */
+static inline void inhibit_xt(void)
+{
+    clock_hi();
+    data_lo();
+}
+#endif
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/ibmpc.h:204]:
```suggestion
```

File: keyboards/mitosis/keymaps/qmitosis/rules.mk [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+#COMMAND_ENABLE = no          # Commands for debug and configuration
+#NKRO_ENABLE = no             # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality
+MIDI_ENABLE = no             # MIDI controls
+AUDIO_ENABLE = no            # Audio output on port C6
+UNICODE_ENABLE = no          # Unicode
+BLUETOOTH_ENABLE = no        # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no         # Enable WS2812 RGB underlight.
+#API_SYSEX_ENABLE = no
+
```
Comment (fauxpark) [inline keyboards/mitosis/keymaps/qmitosis/rules.mk:16]:
```suggestion
MOUSEKEY_ENABLE = no
UNICODE_ENABLE = no
LTO_ENABLE = yes
```

File: keyboards/lily58/keymaps/marfritlily/rules.mk [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+SLEEP_LED_ENABLE = yes    # Breathing sleep LED during USB suspend
+
+# If you want to change the display of OLED, you need to change here
+SRC +=  ./lib/rgb_state_reader.c \
+        ./lib/layer_state_reader.c \
+        ./lib/logo_reader.c \
+        ./lib/keylogger.c \
+        # ./lib/mode_icon_reader.c \
+        # ./lib/host_led_state_reader.c \
+        # ./lib/timelogger.c \
```
Comment (fauxpark) [inline keyboards/lily58/keymaps/marfritlily/rules.mk:31]:
```suggestion
EXTRAKEY_ENABLE = yes
SLEEP_LED_ENABLE = yes
RGBLIGHT_ENABLE = yes
UNICODE_ENABLE = yes
LTO_ENABLE = yes

# If you want to change the display of OLED, you need to change here
SRC +=  ./lib/rgb_state_reader.c \
        ./lib/layer_state_reader.c \
        ./lib/logo_reader.c \
        ./lib/keylogger.c \
        # ./lib/mode_icon_reader.c \
        # ./lib/host_led_state_reader.c \
        # ./lib/timelogger.c \
```

File: keyboards/converter/usb_usb/keymaps/marhasu/rules.mk [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+COMMAND_ENABLE = no          # Commands for debug and configuration
+NKRO_ENABLE = no             # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no        # Enable keyboard backlight functionality
+MIDI_ENABLE = no             # MIDI controls
+AUDIO_ENABLE = no            # Audio output on port C6
+UNICODE_ENABLE = no          # Unicode
+BLUETOOTH_ENABLE = no        # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no         # Enable WS2812 RGB underlight.
+API_SYSEX_ENABLE = no
+
```
Comment (fauxpark) [inline keyboards/converter/usb_usb/keymaps/marhasu/rules.mk:16]:
```suggestion
EXTRAKEY_ENABLE = no
LTO_ENABLE = yes
```

File: keyboards/converter/ibmpc_usb/rules.mk [hunk 19]
```diff
@@ -0,0 +1,28 @@
...
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
+API_SYSEX_ENABLE = no
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+CUSTOM_MATRIX = yes
+
+SRC = matrix.c ibmpc.c
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/rules.mk:28]:
```suggestion
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = yes           # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
CUSTOM_MATRIX = yes

SRC += matrix.c ibmpc.c
```

File: keyboards/converter/ibmpc_usb/keymaps/marfrit/rules.mk [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
+API_SYSEX_ENABLE = no
+G80_2551_SUPPORT = no
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/keymaps/marfrit/rules.mk:21]:
```suggestion
NKRO_ENABLE = no
```

File: keyboards/converter/ibmpc_usb/keymaps/default/rules.mk [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+MIDI_ENABLE = no            # MIDI controls
+AUDIO_ENABLE = no           # Audio output on port C6
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
+API_SYSEX_ENABLE = no
+G80_2551_SUPPORT = no
+
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no    # Breathing sleep LED during USB suspend
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/keymaps/default/rules.mk:21]: This file can be deleted.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/40percentclub/foobar/keymaps/qfoobar/config.h (hunk 6 above)
Comment (marfrit) [reply keyboards/40percentclub/foobar/keymaps/qfoobar/config.h:1]: qfoobar shouldn't be in this, I messed up with my own keymaps, will remove this.

File: keyboards/converter/usb_usb/keymaps/marhasu/rules.mk (hunk 18 above)
Comment (marfrit) [reply keyboards/converter/usb_usb/keymaps/marhasu/rules.mk:16]: will be deleted

File: keyboards/converter/ibm_terminal/keymaps/default/rules.mk [hunk 22]
```diff
@@ -1,20 +1,19 @@
 # Build Options
-#   change to "no" to disable the options, or define them in the Makefile in 
+#   change to "no" to disable the options, or define them in the Makefile in
 #   the appropriate keymap folder that will get included automatically
```
Comment (fauxpark) [inline keyboards/converter/ibm_terminal/keymaps/default/rules.mk:3]:
```suggestion
#   change yes to no to disable
```

File: keyboards/converter/ibm_terminal/keymaps/default/rules.mk [hunk 23]
```diff
@@ -1,20 +1,19 @@
...
 #
 BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
-MOUSEKEY_ENABLE = yes        # Mouse keys
+MOUSEKEY_ENABLE = no        # Mouse keys
```
Comment (fauxpark) [inline keyboards/converter/ibm_terminal/keymaps/default/rules.mk:6]:
```suggestion
MOUSEKEY_ENABLE = no         # Mouse keys
```

File: keyboards/converter/ibm_terminal/keymaps/default/rules.mk [hunk 24]
```diff
@@ -1,20 +1,19 @@
...
-UNICODE_ENABLE = no         # Unicode
-BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
-RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.
-PS2_USE_USART = yes
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+BLUETOOTH_ENABLE = no       # Enable Bluetooth
+AUDIO_ENABLE = no           # Audio output
+PS2_USE_USART = no
+PS2_USE_INT   = yes
 API_SYSEX_ENABLE = no
```
Comment (fauxpark) [inline keyboards/converter/ibm_terminal/keymaps/default/rules.mk:17]:
```suggestion
```

File: keyboards/converter/ibmpc_usb/keymaps/marfrit/rules.mk [hunk 25]
```diff
@@ -0,0 +1,9 @@
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in
+#   the appropriate keymap folder that will get included automatically
+#
```
Comment (fauxpark) [inline keyboards/converter/ibmpc_usb/keymaps/marfrit/rules.mk:4]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

**Review by fauxpark — CHANGES_REQUESTED** [review]
This should really be written in C.

**Comment by marfrit** [comment]
> This should really be written in C.

I removed the C++ stuff for the time being.

**Comment by Muirium** [comment]
Closed? Aw. QMK really needs this.

**Comment by alvinlimfuel** [comment]
QMK has long moved past its vintage roots

**Pull request closed without merging**